### ISSUE TITLE
compilation issues on x86_64 (3.18.2) and ARMv7a (linux-rpi-3.18.y)

### ISSUE THREAD

**Issue by k-a-z-u** [xqms/xtion#9]
I just discovered this great piece of software but had/have issues compiling the module.

On 64-Bit (gentoo) linux with kernel 3.18.2 I get "SSE register return with SSE disabled" errors for parts of xtion-math-emu and xtion-depth due to -mno-sse build behavior for kernel modules. Adding "-msse" to the Makefile, the build succeeds and the module works.
However, if I understand correctly, such floating-point-math (and SSE?) regions must be guarded with kernel_fpu_begin() and kernel_fpu_end(). I guess on 32-Bit it runs out-of-the-box?

When (cross)-compiling the module for the Raspberry Pi 2, I get undefined references to "__aeabi_fmul" and various other soft-float functions.
Adding "-mfloat-abi=softfp" to the Makefile, the build succeeds. Loading the module, however, crashes the kernel with various errors.
But this might as well be an issue with something missing within my Raspberry-Kernel like soft-float emulation or something like that?

Unfortunately I'm fairly new to kernel-modules and hardware-specifics like FPU and SSE. Nevertheless I want to help out to get this neat piece of software up and running ;)

**Comment by k-a-z-u** [comment]
After many hours of reading, guessing, and testing, I think I found the solution.
`xtion_generate_lut()` uses some #define floats that can be just casted to int as they have no decimals? After doing so, one final float operation remains:
```
a = 8 * PARAM_COEFF * SHIFT_SCALE * xtion->fixed.reference_distance;`
aa.uint32_ = u2f(a);
```
I changed this one to
```
reference_distance_f32.float_ = xtion->fixed.reference_distance;
...
aa.uint32_ = u2f(8 * PARAM_COEFF * SHIFT_SCALE);
mul_f32(&reference_distance_f32, &aa, &aa);
mul_f32(&dcmos_emitter_distance_f32, &aa, &aa);
```
and now it compiles with the default Makefile on both x86_64 and the Raspberry Pi2.
The resulting depth image looks still valid on x86_64. Havent tested the image on the Raspberry, but the module now loads without crashing the kernel ;)

If you are interested (and above assumptions were correct), I will combine these changes into a pull-request :)

**Comment by xqms** [comment]
@k-a-z-u Thanks, I'd welcome a PR with your improvements.

Until recently (until 477218bf155c8fae83abe6b50dfee783c398c9ac) we had a static LUT just for xtion devices. That would have been my next suggestion to try on the Raspberry. Fixing the LUT generator is even better ;-)

@arnike Do you have time to check that the changes are okay - also for the PrimeSense devices?

**Comment by arnike** [comment]
@k-a-z-u the floats in the #define's are indeed needless there, thanks for pointing this out.
@xqms Sure, I'll test it on Tuesday when I get to a primesense if there is no rush.

**Comment by arnike** [comment]
Looks fine to me for PrimeSense. I was thinking, if one could also perhaps change the #defines in xtion-depth.h to integers to avoid unnecessary float -> int casts.

**Comment by k-a-z-u** [comment]
I can have a look whether changing the #defines instead of casting is possible (find all usages of the #defines). Maybe we should drop the pending pull-request and I create a new one including the changes from the 2 other issues after they are confirmed/checked :)

**Comment by xqms** [comment]
I merged the #10 as it was, since the defines were already there. Of course I'd welcome a further PR with refactoring.

@arnike: Thanks for checking!

Since the original issue is fixed by #10, I'm closing this ticket now.
